components/constellation/constellation.rs

```diff
@@ -259,6 +259,33 @@ struct BrowsingContextGroup {
     webgpus: HashMap<Host, WebGPU>,
 }
 
+// Helper macro that returns the pipeline for a given browsing context
+// or returns after printing a warning.
+macro_rules! maybe_pipeline {
+    ($this:ident, $bc_id:ident, $warn_msg:expr) => {
+        match $this.browsing_contexts.get(&$bc_id) {
+            Some(ctx) => ctx.pipeline_id,
+            None => return warn!("{}: {}", $bc_id, $warn_msg),
+        }
+    };
+}
+
+// Helper macro that sends a message to the event loop of a given pipeline
+// or returns after printing a warning.
+macro_rules! maybe_send {
+    ($this:ident, $pipeline_id:ident, $msg:expr, $warn_msg:expr) => {
+        let result = match $this.pipelines.get(&$pipeline_id) {
+            Some(pipeline) => pipeline.event_loop.send($msg),
+            None => {
+                return warn!("{}: {}", $pipeline_id, $warn_msg);
+            },
+        };
+        if let Err(e) = result {
+            $this.handle_send_error($pipeline_id, e);
+        }
+    };
+}
+
```

Comment:
I think we should avoid metaprogramming for these type of helpers. Please remove these or convert them into functions. I think also that `maybe_send` isn't very descriptive of what happens in a failure case. I would call them something like "send_handling_error" or something along those lines.

Reply:
In order to do an early return, have the function return `false` on failure and include a return statement. Returning from a function in a macro allows us to remove lines of code, but makes it very hard to follow program flow for anyone reading the code.